### ISSUE TITLE
bug: potential metrics leak

### ISSUE THREAD

**Issue by hzxa21** [risingwavelabs/risingwave#6855]
### Describe the bug

In our system, metrics label are widely used but I nonticed that the labeled metrics are not cleaned when they become unused. There are potential metrics leak for the following metrics:
- Per batch task metrics: should be cleaned when the task finishes.
- Per actor metrics: should be cleaned when the actor is dropped or moved.
- Per source metrics: should be cleaned when the source is dropped.
- Exchange metrics: should be cleaned when the actor is dropped or moved.
- Per table storage metrics: should be cleand when the table id dropped.
- Per host compaction metrics: shold be cleaned when the host is offline.
- Per compaction group metrics ([meta side metrics](https://github.com/risingwavelabs/risingwave/blob/dccb53ed871106aaaffe04ccaf9e4ea2daeeced7/src/meta/src/hummock/metrics_utils.rs#L154) are cleaned up): should be cleaned when the group is dropped.

### Additional context

To fix the issue, we can cache the labeled metrics `with_label_values` and do `remove_label_values`  when the metrics are no longer used. This has a side benefit to reduce label lookup when updating the metrics value.

**Comment by soundOfDestiny** [comment]
> Per compaction group metrics ([meta side metrics](https://github.com/risingwavelabs/risingwave/blob/dccb53ed871106aaaffe04ccaf9e4ea2daeeced7/src/meta/src/hummock/metrics_utils.rs#L154) are cleaned up): should be cleaned when the group is dropped.

now it is dealt by `remove_compaction_group_in_sst_stat` which is called in `sync_group`

**Comment by hzxa21** [comment]
> > Per compaction group metrics ([meta side metrics](https://github.com/risingwavelabs/risingwave/blob/dccb53ed871106aaaffe04ccaf9e4ea2daeeced7/src/meta/src/hummock/metrics_utils.rs#L154) are cleaned up): should be cleaned when the group is dropped.
> 
> now it is dealt by `remove_compaction_group_in_sst_stat` which is called in `sync_group`

Yes, meta side compaction group metrics are already cleaned up correctly but the compactor side are still not.

**Comment by Gun9niR** [comment]
Cleaning the labels right after a short-living instance is dropped/finished will cause the corresponding metric value to be removed before it is collected. For example, a batch task can take only 1 sec, but Prometheus may collect once every 20s. Therefore, the stale labels can only be removed after they are collected, which means we need to implement our own `Collector` to be aware of the invocation of `collect()`. In #5770, I call `reset()` on all the metrics after they are collected, but it has a race condition that if data is written between `collect()` and `reset()`, it will be lost, tho it happens very rarely.

**Comment by lmatz** [comment]
Give an example:

<img width="917" alt="image (2)" src="https://user-images.githubusercontent.com/5791930/226850416-c5a56969-bce0-43c6-8a46-cf97decac087.png">

The top ones come from a streaming query that has been dropped. But the metrics are kept.

**Comment by fuyufjh** [comment]
Need to call `remove_label_values` when actor is dropped.
